test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,422 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.loopopts;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Asserts;
+import jdk.test.lib.Utils;
+import java.util.Random;
+
+/*
+ * @test
+ * @bug 8323220
+ * @summary Test loop invariant code motion for cmp nodes through reassociation
+ * @library /test/lib /
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
+ */
+public class InvariantCodeMotionReassociateCmp {
+    private static final Random RANDOM = Utils.getRandomInstance();
+    private static final int size = 500;
+    private int inv1;
+    private int inv2;
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @DontInline
+    private void blackhole() {}
+
+    @Setup
+    public Object[] setupEq() {
+        inv1 = RANDOM.nextInt();
+        inv2 = inv1 + RANDOM.nextInt(size * 2);
+        return new Object[] { inv1, inv2 };
+    }
+
+    @Setup
+    public Object[] setupNe() {
+        inv1 = RANDOM.nextInt();
+        if (RANDOM.nextInt() % 7 == 0) {
+            if (inv1 == 0) {
+                inv1 = 1;
+            }
+            // Setup inputs to be equals sometimes to avoid uncommon traps
+            inv2 = inv1;
```

Comment:
You may want to use the `SetupInfo` instead, and make this deterministic. Just in the case where the randomness here fails in all cases - then you would not have any equal case.
Also: why can `inv1` not be zero?

Reply:
Sure, can look into `SetupInfo`.

`inv1` can be zero. I'll remove that block.